mbuild/compound.py

```diff
@@ -278,6 +279,140 @@ def _contains_only_ports(self):
+
+        # if index is specified, ensure we are not selecting an index out of range
+        if not index is None:
+            if index >= len(self.children):
+                raise MBuildError(
+                    f"Index {index} out of range. The number of first level nodes in the tree is {len(self.children)}."
+                )
+
+        count = -1
+
+        for h in hh:
+            if len(h["comp"].children) == 0:
+                n_bonds = h["comp"].n_direct_bonds
+            else:
+                n_bonds = h["comp"].n_bonds
+            if h["level"] == 0:
+                count = count + 1
+            if print_full:
+                if index is None:
+                    tree.create_node(
+                        f"[{h['comp'].name}]: {h['comp'].n_particles} particles, {n_bonds} bonds, {len(h['comp'].children)} children",
+                        f"{h['comp_id']}",
+                        f"{h['parent_id']}",
+                    )
+                elif count == index:
+                    tree.create_node(
+                        f"[{h['comp'].name}]: {h['comp'].n_particles} particles, {n_bonds} bonds, {len(h['comp'].children)} children",
+                        f"{h['comp_id']}",
+                        f"{h['parent_id']}",
+                    )
+            else:
+                tree.create_node(
+                    f"[{h['comp'].name} x {h['n_dup']}], {h['comp'].n_particles} particles, {n_bonds} bonds, {len(h['comp'].children)} children",
+                    f"{h['comp_id']}",
+                    f"{h['parent_id']}",
+                )
+        if show_tree:
+            tree.show()
+        return tree
+
+    def _get_hierarchy(self, level=0):
+        """Return an array of dictionaries corresponding to hierarchy of the compound, recursively."""
+        if not self.children:
+            return
+        for child in self.children:
+            yield {
+                "level": level,
+                "parent_id": id(self),
+                "comp_id": id(child),
+                "comp": child,
+            }
+            for subchild in child._get_hierarchy(level + 1):
+                yield subchild
+
+    def _get_hierarchy_nodup(self, level=0):
+        """Return an array of dictionaries corresponding to hierarchy of the compound, recursively.
+        This routine will identify any duplicate compounds at a given level, including the number of
+        duplicates for each compound. Compounds are considered to be identical if the name,
+        number of children, and number of particles are the same at the same level.
+        """
+        if not self.children:
+            return
+
+        duplicates = {}
+        for child in self.children:
+            part_string = "".join([part.name for part in child.particles()])
+            child_string = "".join([child.name for child in child.children])
+            identifier = f"{child.name}_{len(child.children)}_{child_string}_{child.n_particles}_{part_string}"
```

Comment:
Do you think it may be worthwhile to also add bond as part of the identifier? may be just number of bonds

Reply:
Adding in number of bonds and a similar "bond_string" seems like a good idea to ensure uniqueness.

Reply:
Actually, creating a bond string seems like it could be problematic.  We can't guarantee that the bond graph edge necessarily returns the pair in the same order each time (some quick tests confirm this...e.g., creating a system with a few ethanes seems each ethane as unique and the underlying CH3 molecules often to be unique...and I get different behaviors if I reinitialize it each time,  where I'm cloning an ethane ).    I don't think we need to add in a bunch of extra code to compare the set of bonds.  I think the follow should be more than sufficient to uniquely identify compounds. 

- compound name
- number of children
- name of children
- number of particles
- name of particles
- number of bonds

an alternate approach would be to convert the compound to a networkx graph and just compare graphs but that might be overkill for this. It would also require some substantial changing the networkx conversion code already in there.